specs/WebMessageObjects.md

```diff
@@ -0,0 +1,355 @@
+# Background
+This is a proposal for a new API that will provide a framework for native representation of DOM
+objects from page content in WebView2 and its implementation for DOM File objects.. The (specific
+ask from WebView2)[https://github.com/MicrosoftEdge/WebView2Feedback/issues/501] is to be able get
+to the paths for DOM file objects, which is not accessible to page content. We also have asks to
+expose other DOM objects, including iframes, <object> objects, etc. which will be added under this
+WebMessageObjects framework in future.
+
+Also in future we want to be able to to inject DOM objects into WebView2 content constructed via the
+app and via the CoreWebView2.PostWebMessage API in the other direction. This API surface needs to be
+compatible with that. (See Appendix)
+
+# Conceptual pages (How To)
+WebMessageObjects are representations of DOM objects that can be passed via the [WebView2 WebMessage
+API](https://learn.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.core.corewebview2webmessagereceivedeventargs).
+You can examine supported DOM objects using native reflections of the types.
+
+Currently the only supported DOM object type with this API is:
+- [File](https://developer.mozilla.org/en-US/docs/Web/API/File)
+
+Page content in WebView2 can pass objects to the app via the
+`chrome.webview.postMessageWithAdditionalObjects(string message, array<object> objects)` content API
+that takes in the array of such supported DOM objects or you can also use
+(ExecuteScript)[https://learn.microsoft.com/en-us/dotnet/api/microsoft.web.webview2.winforms.webview2.executescriptasync]
+with same API. If an invalid or unsupported object is passed via this API, an exception will be

```

Comment:
How do I know which objects are supported?  Do I try sending all the objects I want to send and prepare for an exception if the version of the webview extending EventTarget doesn't allow it?  (It isn't clear to me whether it is "caller error" to, in the future, be passing through object types that got later support. So I'm not sure if rejecting AdditionalObjects is an exceptional case.)

Reply:
Native side can do feature detection and inform script side. So leave as is (but also we'll look into how window.postMessage handles this for its transferable array.)